examples/ahfe.py

```diff
@@ -35,46 +44,58 @@ def get_romol_conf(mol):
+
+combined_masses = np.concatenate([host_masses, ligand_masses])
 
 # write some conformations into this PDB file
-writer = pdb_writer.PDBWriter([topology, romol], "debug.pdb")
+writer = pdb_writer.PDBWriter([omm_topology, romol], "debug.pdb")
 
 # note the order in which the coordinates are concatenated in this step --
 #   in a later step we will need to combine recipes in the same order
-combined_conf = np.concatenate([host_coords, ligand_coords])
+combined_coords = np.concatenate([host_coords, ligand_coords])
 
 num_host_atoms = host_coords.shape[0]
 
 # note: .py file rather than .offxml file
 # note: _ccc suffix means "correctable charge corrections"
 ff_handlers = deserialize_handlers(open('ff/params/smirnoff_1_1_0_ccc.py').read())
 
-ligand_recipe = Recipe.from_rdkit(romol, ff_handlers)
+final_potentials = []
+final_vjp_and_handles = []
 
-# note: "bound" potentials are energy functions where the parameters have been
-#   set to a constant using .bind()
-#   (contrast with "unbound" potentials, for cases where the parameters are
-#   variable and where we may be interested in dU/dparams)
-for bp in ligand_recipe.bound_potentials:
+# keep the bonded terms in the host the same.
+# but we keep the nonbonded term for a subsequent modification
+for bp in host_bps:
     if isinstance(bp, potentials.Nonbonded):
-        # note: this is something like "bp.set_lambda_offset_idxs(ones)"
-        #   (at initialization, offset_indices are all zeros)
-        array = bp.get_lambda_offset_idxs()
-        array[:] = 1 
-
-water_recipe = Recipe.from_openmm(system)
-
-# note: this step is dependent on order -- ensure consistency with the order
-#   in which coordinates were concatenated above
-# note: calling `recipe.combine(other)` will offset all of the bond indices etc.
-#   within the sub-recipe described by `other`
-combined_recipe = water_recipe.combine(ligand_recipe)
-
-# note: lambda goes from 0 to +infinity, where we approximate +infinity with
-#   any number of nanometers sufficiently large that the ligand has almost no
-#   interaction with the rest of the system (in this case, 1.2 nm will do)
-for final_lamb in np.linspace(0, 1.2, 8):
+        host_p = bp
+    else:
+        final_potentials.append(bp)
+        final_vjp_and_handles.append(None)
+
+ff = Forcefield(ff_handlers)
+gbt = topology.BaseTopology(romol, ff)
+hgt = topology.HostGuestTopology(gbt, host_p, num_host_atoms)
+
+# setup the parameter handlers for the ligand
+tuples = [
+    [hgt.parameterize_harmonic_bond, [ff.hb_handle]],
+    [hgt.parameterize_harmonic_angle, [ff.ha_handle]],
+    [hgt.parameterize_proper_torsion, [ff.pt_handle]],
+    [hgt.parameterize_improper_torsion, [ff.it_handle]],
+    [hgt.parameterize_nonbonded, [ff.q_handle, ff.lj_handle]],
+]
+
+# instantiate the vjps while parameterizing (forward pass)
+for fn, handles in tuples:
+    params, vjp_fn, potential = jax.vjp(fn, *[h.params for h in handles], has_aux=True)
```

Comment:
@maxentile this is the glorious part! vjps are computed automagically here.

The hgt.parameterize_nonbonded function takes in ligand forcefield parameters, and emits system parameters. Its vjp is now computed with jax. The vjp_fn takes in system parameter adjoints and returns forcefield parameter adjoints!